Refactor nibbles

David-Petrov opened the pull request:
refactor(nibbles)!: 

* make Nibbles immutable
* turn constructors into public factory methods for Nibbles that copy data on construction to gain ownership, instead leaving the constructor package protected in order to still make us of internal list's mutability when modifying the Nibbles.
* `Nibbles implements Iterable<Nibble>` instead of Collection or List
* remove generic partial key type argument from `DecodedNode` since in Java's eyes `Nibbles` is no more a collection, thus we can't access its `size()` method generically. Instead, we simply specify that generic type to Nibbles
* make `NibblesToBytes` a static utility class

Commit:
chore: clean up BytesToNibbles

Commit:
refactor(nibbles)!:

* make Nibbles immutable

* make public factory methods for Nibbles that copy data on construction to gain ownership

* `Nibbles implements Iterable<Nibble>` instead of Collection or List

* remove generic partial key type argument from `DecodedNode` since in Java's eyes `Nibbles` is no more a collection, thus we can't access its `size()` method generically. Instead, we simply specify that generic type to Nibbles

* make `NibblesToBytes` a static utility class

David-Petrov commented on src/main/java/com/limechain/trie/structure/nibble/Nibbles.java:
```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
+     * @return the new Nibbles after insertion
+     * @implNote will create a new modified copy of the nibbles, does not mutate the instance invoked on
+     */
+    public Nibbles add(Nibble nibble) {
+        List<Nibble> newNibbles = new ArrayList<>(this.nibbles);
+        newNibbles.add(nibble);
+        return new Nibbles(newNibbles);
     }
```
I don't like how mutable this method name sounds, but I decided it should be named after what it actually does and not after how it certainly behaves. Resolve if you don't take issue in that.

David-Petrov commented on src/main/java/com/limechain/trie/structure/nibble/Nibbles.java:
```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
-    @Override
-    public Nibble set(int index, Nibble element) {
-        return nibbles.get(index);
+    // NOTE:
+    //  Since Nibbles is immutable, we don't actually need to explicitly copy
+    //  This method could go away, passing by reference could suffice, although going back to mutability
+    //  if later decided would become rather difficult.
```
That's an interesting conundrum, take a look at the note.

georg-getz replied:
Why would mutability become difficult to switch to?

David-Petrov replied:
Because someone would have to track the ownership anew, which is currently given by `copy`. Everywhere we now copy, the new owner would want to own the data and probably mutate it at some point. These mutations generate new instances now, which is the easy part to refactor (when switching to mutability of Nibbles) - make them mutate the old instance, but that would cause changes in previous owners too, since by removing `copy` we will have passed by reference only.

David-Petrov replied:
We decided we'll leave `copy` as a shallow implementation returning `this` only for explicitness and traceability of ownership's sake.

David-Petrov commented on src/main/java/com/limechain/trie/structure/nibble/Nibbles.java:
```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
+    public Nibbles drop(int n) {
+        // NOTE: We might want to not throw, but return an empty Nibbles instead if (n < 0 || n > size).
+        return new Nibbles(this.nibbles.subList(n, this.nibbles.size()));
     }
 
-    @Override
-    public Nibble remove(int index) {
-        return nibbles.remove(index);
+    /**
+     * Takes only the first n nibbles from the beginning.
+     * @param n the number of nibbles to take
+     * @return a new Nibbles limited to only the first n leading nibbles
+     * @throws IndexOutOfBoundsException if
+     *         ({@code n < 0 || n > size})
+     */
+    public Nibbles take(int n) {
+        // NOTE: We might want to not throw, but return the same Nibbles instead if (n > size).
+        return new Nibbles(this.nibbles.subList(0, n));
     }
```
I couldn't decide on the behaviour I prefer for these two (although it doesn't matter as much, but still).
Since these methods come from rather abundant outside usage of `.stream().skip(n).collect(NibblesCollector.toNibbles())`, it makes sense for them to behave like `Stream`'s skip / take (hence not throwing anything). 
On the other hand, turning those into methods of Nibbles allows us to make use of the underlying list and only copy a view of the sublist we're interested in and simply wrap it into a new nibbles (without unnecessary copying). But `subList` throws exceptions, so it made natural sense to do this as well.

Ultimately, we could add a simple check for the boundaries and act stream-likey while still taking advantage of `subList`.

georg-getz replied:
I'm fine with them throwing exceptions.

David-Petrov commented on src/main/java/com/limechain/trie/structure/nibble/Nibbles.java:
```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
-    @Override
-    public boolean removeAll(@NotNull Collection<?> c) {
-        return nibbles.removeAll(c);
+    private static String toLowerHexString(Iterable<Nibble> nibbles) {
```
I don't like this method being in here, but couldn't figure out a better place. If you think it's ok, let it be.

stoyan-lime replied:
You can move it to a utility class for nibble operations (NibbleConverter or something similar?). You can also move the methods from NibblesToBytes there.

georg-getz replied:
We decided that `BytesToNibbles` and the reverse one could be in a single utility class where it would also make sense to put this.

David-Petrov replied:
Take a look and tell me whether you like it: a3f60bfc6ac55054ac4f19506131a7652eca6ca9
Resolve if considered finished.

bokoto000 commented on src/main/java/com/limechain/trie/structure/nibble/Nibbles.java:
```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
+     * @throws IndexOutOfBoundsException if
+     *         ({@code n < 0 || n > size})
+     */
+    public Nibbles drop(int n) {
```
I smell some Haskell-ish naming here. Would subNibbles(int from) sound better?

David-Petrov replied:
I thought about something similar to Java's `subList` naming, but honestly, I'd prefer the more explicit verbs. Stream API's `skip` and `limit` were also considered, but they seem kind of arbitrary to me (I don't like `limit` but I can see where the Java authors have come from). Initially, they were `take` and `skip`, but then `drop` clicked more with me 😄 
Feel free to discuss it with the team, ultimately I'm up for anything.

bokoto000 commented on src/main/java/com/limechain/trie/structure/nibble/Nibbles.java:
```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
+     * @throws IndexOutOfBoundsException if
+     *         ({@code n < 0 || n > size})
+     */
+    public Nibbles take(int n) {
```
Again smells like Haskell. Would subNibbles(int to) sound more Java-ish?

stoyan-lime commented on src/main/java/com/limechain/trie/structure/nibble/Nibbles.java:
```diff
@@ -1,86 +1,78 @@
 
-    public static Nibbles of(int... nibbles) {
-        return new Nibbles(Arrays.stream(nibbles).mapToObj(Nibble::fromInt));
+    Nibbles(List<Nibble> nibbles) {
```
You can make the constructor private and create a public `of(List)` method to keep the class consistent.

Commit:
chore: remove resolved note

Commit:
refactor: reimplement `copy` as a shallow marker method

also, add javadoc

Commit:
refactor: extract static methods into `NibblesUtils`

thus fusing `NibblesToBytes` and `toLowerHexString` together

Commit:
chore: remove code smell

mark `NibblesUtils` as a lombok `@UtilityClass` thus removing the need to manually handle constructors and such stuff

Commit:
refactor!: optimize bytes to nibbles conversion

remove the unnecessary copying before joining nibbles to form bytes, by refactoring the implementation of the public methods
